README.md
# AliYun NLS

[![js-standard-style](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://github.com/feross/standard)
[![action][ci-image]][ci-url]
[![license][license-image]][repository-url]
[![NPM version][npm-image]][npm-url]
[![npm download][download-image]][download-url]

[ci-image]: https://img.shields.io/github/workflow/status/funnyzak/aliyun-nls/Node.js%20CI
[ci-url]: https://github.com/funnyzak/aliyun-nls/actions
[license-image]: https://img.shields.io/github/license/funnyzak/aliyun-nls.svg?style=flat-square
[repository-url]: https://github.com/funnyzak/aliyun-nls
[npm-image]: https://img.shields.io/npm/v/aliyun-nls.svg?style=flat-square
[npm-url]: https://npmjs.org/package/aliyun-nls
[download-image]: https://img.shields.io/npm/dm/aliyun-nls.svg?style=flat-square
[download-url]: https://npmjs.org/package/aliyun-nls

阿里云智能语音处理 Node 模块。

## 目录

- [AliYun NLS](#aliyun-nls)
  - [目录](#目录)
  - [开始](#开始)
  - [长文本语音合成](#长文本语音合成)
    - [用例](#用例)
    - [函数](#函数)
      - [`checkConfig(): Promise<boolean>`](#checkconfig-promiseboolean)
      - [`task(text: string, options?: AliyunTTS.TTSOption): Promise<string>`](#tasktext-string-options-aliyunttsttsoption-promisestring)
      - [`status(taskId: string, appKey?: string): Promise<AliyunTTS.TTSComplete>`](#statustaskid-string-appkey-string-promisealiyunttsttscomplete)
      - [`taskSync(text: string, options: AliyunTTS.TTSOption, interval?: number): Promise<AliyunTTS.TTSComplete>;`](#tasksynctext-string-options-aliyunttsttsoption-interval-number-promisealiyunttsttscomplete)
    - [定义](#定义)
      - [AliyunTTS.TTSOption](#aliyunttsttsoption)
      - [AliyunTTS.TTSComplete](#aliyunttsttscomplete)
  - [Author](#author)
  - [参考](#参考)
  - [License](#license)

## 开始

from [npm](https://github.com/npm/npm)

    $ npm install/aliyun-nls

## 长文本语音合成

### 用例

```js
const { AliyunTTS } = require(/aliyun-nls');

!(async () => {
  const _aliyunTTS = new AliyunTTS(
    {
      accessKeyId: 'accessKeyId',
      accessKeySecret: 'accessKeySecret',
      endpoint: 'https://nls-meta.cn-shanghai.aliyuncs.com',
      nlsUrl: 'https://nls-gateway.cn-shanghai.aliyuncs.com/rest/v1/tts/async',
      apiVersion: '2019-02-28'
    },
    'app key' /** optional **/
  );

  // test aliyun api config
  console.log((await _aliyunTTS.checkConfig()) ? 'passed' : 'error');

  // more ...
})();
```

了解更多 [TTS Define](https://github.com/funnyzak/aliyun-nls/blob/master/nls.d.ts).

### 函数

`AliyunTTS` 有如下方法:

#### `checkConfig(): Promise<boolean>`

返回值 `Promise<boolean>` - 返回阿里云语音配置密钥是否有效。

```js
const checkRlt = await _aliyunTTS.checkConfig();
console.log(checkRlt ? 'the config is passed' : 'error config');
```

#### `task(text: string, options?: AliyunTTS.TTSOption): Promise<string>`

- `text` string - 要转换的文本。
- `options` NLSOption (optional) - 高级设置。
- `appKey` string (optional) - 应用 key。

返回值 `Promise<string>` - 返回转换任务 ID。

#### `status(taskId: string, appKey?: string): Promise<AliyunTTS.TTSComplete>`

查询转换状态。

- `taskId` string - 任务 ID。
- `appKey` string (optional) - 应用 Key。

返回值 `Promise<AliyunTTS.TTSComplete>` - 转换状态。

```js
const taskId = await _aliyunTTS.task('你好，世界！', {
  /** options **/
});

console.log(taskId);
```

#### `taskSync(text: string, options: AliyunTTS.TTSOption, interval?: number): Promise<AliyunTTS.TTSComplete>;`

同步返回合成结果。

- `text` string - 要转换的文本。
- `options` NLSOption - 高级设置。
- `interval` number (optional) - 检查转换状态的轮训事件间隔（秒）。

返回值 `Promise<AliyunTTS.TTSComplete>` - 转换状态。

```js
const rlt = await _aliyunTTS.taskSync('你好，世界！', {
  format: 'mp3'
  sample_rate: 16000,
  voice: 'xiaoyun',
  volume: 50,
  speech_rate: 50,
  pitch_rate: 50
  /* options */
});

console.log('complete result => ', JSON.stringify(rlt));
```

### 定义

#### AliyunTTS.TTSOption

- `appKey` - 应用 Key,可选。
- `format` - 音频编码格式，支持 pcm/wav/mp3 格式，默认是 pcm。
- `sample_rate` - 音频采样率，支持 16000Hz 和 8000Hz，默认是 16000Hz。
- `voice` - 发音人，默认是 xiaoyun。更多发音人请参见[接口说明](https://help.aliyun.com/document_detail/130509.htm?spm=a2c4g.11186623.0.0.442a38adeflvK0#topic-2606811)。
- `volume` - 音量，范围是 0~100，默认 50。
- `speech_rate` - 语速，范围是 0-100，默认是 50。
- `pitch_rate` - 语调，范围是 0-100，默认是 50。
- `enable_subtitle` - 是否启用句级时间戳功能，默认值为 false。
- `enable_notify` - 是否启用回调功能，默认值为 false。
- `notify_url` - 回调服务的地址。当 enable_notify 取值为 true 时，本字段必填。URL 支持 HTTP/HTTPS 协议，Host 不能使用 IP 地址。

#### AliyunTTS.TTSComplete

- `task_id` string - 任务 ID。
- `audio_address` string - 合成的音频 URL。
- `notify_custom` string - 回调地址。
- `sentences` json - 句级时间戳对象。
- `appKey`: string - 使用的 APP Key。
- `options`: AliyunTTS.TTSOption - 合成选项。
- `text`: string - 待合成文本。
- `startTime`: number - 合成开始时间（毫秒时间戳）。
- `elapsed`: number - 合成耗时（毫秒）。

## Author

| [![twitter/funnyzak](https://s.gravatar.com/avatar/c2437e240644b1317a4a356c6d6253ee?s=70)](https://twitter.com/funnyzak 'Follow @funnyzak on Twitter') |
| ------------------------------------------------------------------------------------------------------------------------------------------------------ |
| [funnyzak](https://yycc.me/)                                                                                                                           |

## 参考

- [长文本语音合成文档](https://help.aliyun.com/document_detail/130509.htm?spm=a2c4g.11186623.0.0.442a38adeflvK0#topic-2606811)

## License

Apache-2.0 License © 2021 [funnyzak](https://github.com/funnyzak)


CHANGELOG.md
## 0.1.0 (2021-11-20)

Feature:

- 首次发布


src/tts.js
'use strict';

const assert = require('assert');
const AliyunRPC = require('@alicloud/pop-core');
const RP = require('request-promise');
const chalk = require('chalk');

class AliyunNLS {
  constructor(config, appKey = '') {
    assert(config, 'must pass "config"');
    assert(config.accessKeyId, 'must pass "config.accessKeyId"');
    assert(config.accessKeySecret, 'must pass "config.accessKeySecret"');
    assert(config.nlsUrl, 'must pass "config.nlsUrl"');

    this.appKey = appKey;
    this.config = {
      endpoint: 'https://nls-meta.cn-shanghai.aliyuncs.com',
      apiVersion: '2019-02-28',
      ...config
    };
    this.client = new AliyunRPC(this.config);

    this.tokenExpire =
      parseInt((new Date().getTime() / 1000).toString(), 10) - 10;
    this.token = '';

    this.taskMap = {};
  }

  log(msg, level = 'info') {
    const message = typeof msg !== 'string' ? JSON.stringify(msg) : msg;

    const warning = (_message) => chalk`{yellow WARNING:} ${_message}`;
    const info = (_message) => chalk`{magenta INFO:} ${_message}`;
    const error = (_message) => chalk`{red ERROR:} ${_message}`;
    switch (level) {
      case 'error':
        console.info(error(message));
        break;
      case 'warn':
        console.warn(warning(message));
        break;
      default:
        console.info(info(message));
        break;
    }
  }
  /**
   * 获取token
   * @returns
   */
  getToken() {
    return new Promise((resolve, reject) => {
      if (
        this.token.length > 0 &&
        this.tokenExpire * 1000 > new Date().getTime()
      ) {
        resolve(this.token);
      } else {
        this.client
          .request('CreateToken', { Format: 'JSON' })
          .then((res) => {
            if (res.ErrMsg === '') {
              this.tokenExpire = res.Token.ExpireTime;
              this.token = res.Token.Id;
              this.log(`created token: ${JSON.stringify(res)}.`);
              resolve(res.Token.Id);
            } else {
              this.log(`create token fail, message: ${res.ErrMsg}.`, 'error');
              reject(new Error(res.ErrMsg));
            }
          })
          .catch((err) => {
            this.log(`create token error, message: ${err.message}.`, 'error');
            reject(err);
          });
      }
    });
  }

  /**
   * 合成任务，并获取任务ID
   * @param text 文本内容
   * @param options 转换选项
   * @returns
   */
  task(text, options, appKey) {
    return new Promise(async (resolve, reject) => {
      if (!text || text === null || text.length === 0) {
        this.log(`nls text cant's null.`, 'warn');
        reject(new Error('please set text.'));
        return;
      }
      let _token = '';
      try {
        _token = await this.getToken();
      } catch (err) {
        reject(err);
        return;
      }

      const _options = {
        format: 'mp3',
        sample_rate: 16000,
        voice: 'xiaoyun',
        volume: 50,
        enable_notify: false,
        speech_rate: 50,
        pitch_rate: 50,
        ...options
      };

      const requestConfig = {
        method: 'POST',
        uri: this.config.nlsUrl,
        json: true,
        headers: {
          'Content-Type': 'application/json'
        },
        body: {
          payload: {
            tts_request: {
              text,
              format: _options.format,
              sample_rate: _options.sample_rate,
              voice: _options.voice,
              volume: _options.volume,
              speech_rate: _options.speech_rate
                ? (_options.speech_rate - 50) * 10
                : 50, // 0-100 to -500-50 format
              pitch_rate: _options.pitch_rate
                ? (_options.pitch_rate - 50) * 10
                : 50,
              enable_subtitle: _options.enable_subtitle
            },
            enable_notify: _options.enable_notify,
            notify_url: _options.notify_url
          },
          context: {
            device_id: (Math.random() + 1).toString(36).substring(7)
          },
          header: {
            appkey: appKey || this.appKey,
            token: _token
          }
        }
      };

      this.log(`request config: ${JSON.stringify(requestConfig)}`);
      const _startTime = new Date().getTime();

      RP(requestConfig)
        .then((_rlt) => {
          this.log(`get task: ${JSON.stringify(_rlt)}`);

          if (_rlt.data.task_id) {
            this.taskMap[_rlt.data.task_id] = {
              options: _options,
              text: text,
              startTime: _startTime
            };

            resolve(_rlt.data.task_id);
          } else {
            reject(new Error('requested and get task id is null.'));
          }
        })
        .catch((_err) => {
          this.log(_err.message, 'error');
          reject(_err);
        });
    });
  }

  /**
   * 获取转换状态
   * @param taskId
   * @param appKey
   * @returns
   */
  status(taskId, appKey) {
    return new Promise(async (resolve, reject) => {
      let _token = '';
      try {
        _token = await this.getToken();
      } catch (err) {
        reject(err);
        return;
      }
      const _appKey = appKey || this.appKey;
      const _config = {
        method: 'GET',
        uri: `${this.config.nlsUrl}?appkey=${_appKey}&task_id=${taskId}&token=${_token}`,
        json: true
      };

      RP(_config)
        .then((rlt) => {
          this.log(`task status: ${JSON.stringify(rlt)}.`);
          if (rlt.error_code !== 20000000) {
            reject(new Error(rlt.error_message));
          } else {
            const rltData = rlt.data;
            const isComplete =
              rltData.audio_address !== null &&
              rltData.audio_address.length > 0;
            resolve({
              ...rlt.data,
              ...this.taskMap[taskId],
              appKey: _appKey,
              elapsed: isComplete
                ? new Date().getTime() - this.taskMap[taskId].startTime
                : -1
            });
          }
        })
        .catch((err) => {
          this.log(`request task status error: ${err.message}.`, 'error');
          reject(err);
        });
    });
  }

  /**
   * 同步完成转换
   * @param text 文字
   * @param options 转换配置
   * @param 轮训时间 秒
   */
  taskSync(text, options = {}, interval = 2) {
    return new Promise(async (resolve, reject) => {
      let taskId = '';
      try {
        taskId = await this.task(text, options);
      } catch (err) {
        reject(err);
        return;
      }

      const _interval = setInterval(async () => {
        try {
          let rlt;
          try {
            rlt = await this.status(taskId);
          } catch (err) {
            clearInterval(_interval);
            reject(err);
            return;
          }
          if (rlt.audio_address !== null && rlt.audio_address.length > 0) {
            clearInterval(_interval);
            resolve(rlt);
          }
        } catch (err) {
          clearInterval(_interval);
          reject(err);
        }
      }, interval * 1000);
    });
  }

  /**
   *检查配置
   * @returns
   */
  async checkConfig() {
    try {
      await this.getToken();
      return true;
    } catch (err) {
      return false;
    }
  }
}

module.exports = AliyunNLS;
